### PR TITLE
Add ement

### PR DESCRIPTION
**Pull request by progfolio**
### Brief summary of what the package does

Matrix chat client for Emacs.

### Direct link to the package repository

https://github.com/alphapapa/ement.el

### Checklist

Assume you're working on `mpc` mode:

- [x] byte-compiles cleanly
- [x] `M-x checkdoc` is happy. Don't manually write `(provide 'evil-collection-mpc)`, `M-x checkdoc` can do it automatically for you
- [x] define `evil-collection-mpc-setup` with `defun`
- [x] define `evil-collection-mpc-mode-maps` with `defconst`
- [x] All functions should start with `evil-collection-mpc-`

Still a work in progress. I'm not married to any of the bindings, either. Figured I'd submit here so other Ement.el users could comment more easily.

### CODE REVIEW

File: modes/ement/evil-collection-ement.el [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+;;; Code:
+(require 'evil-collection)
+(require 'ement nil t)
+(require 'ement-room-list nil t)
+
+(defgroup evil-collection-ement nil
+  "Evil bindings for Ement.el."
+  :group 'evil-collection)
+
+(defcustom evil-collection-ement-want-auto-retro t
```
Comment (Stebalien) [inline modes/ement/evil-collection-ement.el:34]: Given that this is defined as an advice, this is a bit invasive to enable by default. I'd suggest either just defining/binding wrapper commands, or disabling this by default.

File: modes/ement/evil-collection-ement.el [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    (kbd "g]") 'ement-room-goto-next
+    (kbd "g[") 'ement-room-goto-prev
+    (kbd "gr") 'ement-room-sync
+    (kbd "gu") 'ement-room-goto-fully-read-marker
+    (kbd "i")  'ement-room-send-message
+    (kbd "I")  'ement-room-compose-message
+    (kbd "q")  'quit-window
+    (kbd "S-RET") 'ement-room-compose-send)
+
+  (when evil-collection-ement-want-auto-retro
```
Comment (Stebalien) [inline modes/ement/evil-collection-ement.el:61]: What about `C-u`?

**Review by Stebalien — COMMENTED** [review]
90% subjective review (from a user, not a maintainer)

File: modes/ement/evil-collection-ement.el (hunk 2 above)
Comment (progfolio) [reply modes/ement/evil-collection-ement.el:61]: Not sure what you mean here.

File: modes/ement/evil-collection-ement.el [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  (evil-collection-define-key 'normal 'ement-describe-room-mode-map
+    (kbd "q") 'quit-window)
+
+  (evil-collection-define-key 'normal 'ement-room-list-mode-map
+    (kbd "RET") 'ement-room-list-RET
+    (kbd "<ret>") 'ement-room-list-RET)
+
+  (evil-collection-define-key '(normal motion) 'ement-room-mode-map
+    (kbd "dd") 'ement-room-delete-message
+    (kbd "D")  'ement-room-delete-message
```
[the remark this answers is not in the document]
Comment (progfolio) [reply modes/ement/evil-collection-ement.el:50]: "d" by itself usually requires an object to delete.
I suppose we could skip that (especially considering there's a confirmation prompt).
I'll let others chime in.

File: modes/ement/evil-collection-ement.el (hunk 1 above)
Comment (progfolio) [reply modes/ement/evil-collection-ement.el:34]: Fair point. I'll disable by default.

File: modes/ement/evil-collection-ement.el (hunk 2 above)
Comment (Stebalien) [reply modes/ement/evil-collection-ement.el:61]: I mean, this covers `k` (line up) but not `C-u` (scroll up).

File: modes/ement/evil-collection-ement.el [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    (kbd "a")  'ement-room-transient
+    (kbd "g]") 'ement-room-goto-next
+    (kbd "g[") 'ement-room-goto-prev
+    (kbd "gr") 'ement-room-sync
+    (kbd "gu") 'ement-room-goto-fully-read-marker
+    (kbd "i")  'ement-room-send-message
+    (kbd "I")  'ement-room-compose-message
```
[the remark this answers is not in the document]
Comment (progfolio) [reply modes/ement/evil-collection-ement.el:69]: I'll let others comment on this. I don't have a strong preference.

File: modes/ement/evil-collection-ement.el (hunk 1 above)
Comment (progfolio) [reply modes/ement/evil-collection-ement.el:34]: Implemented via hooks now. Disabled by default.

File: modes/ement/evil-collection-ement.el (hunk 4 above)
[the remark this answers is not in the document]
Comment (alphapapa) [reply modes/ement/evil-collection-ement.el:69]: I'm not even a Vim user, but even I know that to "append" a new message should be `a`.  ;)  (j/k, I don't use EVIL so I don't care)

[the remark this answers is not in the document]
Comment (progfolio) [reply modes/ement/evil-collection-ement.el:69]: > I'm not even a Vim user, but even I know that to "append" a new message should be `a`. ;) (j/k, I don't use EVIL so I don't care)

Makes more sense to me than "c".

File: modes/ement/evil-collection-ement.el [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+(defun evil-collection-ement-setup ()
+  "Set up `evil' bindings for Ement.el."
+  (evil-collection-define-key 'normal 'ement-describe-room-mode-map
+    (kbd "q") 'quit-window)
+
+  (evil-collection-define-key 'normal 'ement-room-list-mode-map
+    (kbd "RET") 'ement-room-list-RET
+    (kbd "<ret>") 'ement-room-list-RET)
+
+  (evil-collection-define-key '(normal motion) 'ement-room-mode-map
```
[the remark this answers is not in the document]
Comment (progfolio) [reply modes/ement/evil-collection-ement.el:64]: Added send + view event bindings.

File: modes/ement/evil-collection-ement.el (hunk 4 above)
[the remark this answers is not in the document]
Comment (progfolio) [reply modes/ement/evil-collection-ement.el]: Moved ement-room-send-message to "a", compose message to "A".

**Comment by edgar-vincent** [comment]
Thanks a lot for doing this @progfolio. I've had a draft in the works for a long time, but never finished it, and I'm glad someone has submitted a PR.

About the bindings themselves, why not simply bind `ement-send-message` to `RET`? I may be mistaken, but it sounds to me like the most obvious choice.

**Comment by progfolio** [comment]
> Thanks a lot for doing this @progfolio. I've had a draft in the works for a long time, but never finished it, and I'm glad someone has submitted a PR.

You're welcome.

> About the bindings themselves, why not simply bind `ement-send-message` to `RET`? I may be mistaken, but it sounds to me like the most obvious choice.

I've added RET and M-RET bindings.

File: modes/ement/evil-collection-ement.el [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+  (evil-collection-define-key 'normal 'ement-room-list-mode-map
+    (kbd "d")     'ement-room-list-kill-buffer
+    (kbd "RET")   'ement-room-list-RET
+    (kbd "<return>") 'ement-room-list-RET)
+
+  (evil-collection-define-key '(normal motion) 'ement-room-mode-map
+    (kbd "<")  'ement-room-transient
+    (kbd "<return>")   'ement-room-send-message
+    (kbd "RET")        'ement-rooom-send-message
```
Comment (daviwil) [inline modes/ement/evil-collection-ement.el:73]: @progfolio Just noticed there is a typo here: `rooom`

Comment (alphapapa) [reply modes/ement/evil-collection-ement.el:73]: Maybe he wanted that command to send a message faster, like "vroooom"...

Comment (daviwil) [reply modes/ement/evil-collection-ement.el:73]: Time to add it!

Comment (progfolio) [reply modes/ement/evil-collection-ement.el:73]: Ooops! Thank yooou.

**Pull request merged**